No-UDP Endpoints: manual data I/O routing

dr7ana opened the pull request:
- Now you can provide a callback that bypasses sending packets through the UDP socket
- This also disables creation of the UDP socket entirely
- The application will need to take responsibility for passing packets into the Endpoint via Endpoint::manually_receive_packet(...)

dr7ana committed:
No-UDP Endpoints: manual data I/O routing

- Now you can provide a callback that bypasses sending packets through the UDP socket
- This also disables creation of the UDP socket entirely
- The application will need to take responsibility for passing packets into the Endpoint via Endpoint::manually_receive_packet(...)
- Note: BOTH endpoints need to be manually routing traffic to one another

jagerman commented on include/oxen/quic/opt.hpp:
```diff
@@ -6,121 +6,157 @@
+    class Endpoint;
 
-    struct max_streams
+    namespace opt
```
The extra indentation here is a bit annoying; perhaps this could be:
```C++

namespace oxen::quic { class Endpoint; }
namespace oxen::quic::opt
{
...
}
```
to keep it less annoying.

dr7ana committed:
011 tunneling test case

jagerman committed:
Use a variant inside Packet instead

Avoids the need for user-defined copy/move constructors.

jagerman commented on include/oxen/quic/endpoint.hpp:
```diff
@@ -88,7 +88,7 @@ namespace oxen::quic
             auto f = p.get_future();
 
-            if (!remote.is_addressable())
+            if (not _manual_routing and !remote.is_addressable())
```
```suggestion
            if (!_manual_routing && !remote.is_addressable())
```
The mixed keyword and non-keyword operators are jarring (and currently everything else in libquic uses symbols instead of keywords).

Side note: I looked for a way to get clang-format to enforce this, but couldn't find one.

jagerman commented on src/udp.cpp:
```diff
@@ -237,7 +237,18 @@ namespace oxen::quic
+#ifndef NDEBUG
+                                    .error_code,
+                            "udp::receive()")
+#endif
```
I don't think this should be `check_rv`'ed for debug builds: it will throw upon failure, which makes the debug and non-debug paths through here substantially different.

jagerman replied:
Checking the return value and logging is fine, just not via `check_rv` because of that throwing behaviour.

jagerman commented on src/udp.cpp:
```diff
@@ -508,14 +519,20 @@ namespace oxen::quic
                 source_addr.v4.ipi_spec_dst
 #endif
                         = path.local.in4().sin_addr;
+                source_addr.v4.ipi_ifindex = 0;
```
Not needed: this will already be 0 via the memset.

jagerman commented on src/udp.cpp:
```diff
@@ -508,14 +519,20 @@ namespace oxen::quic
             else
+            {
                 source_addr.v6.ipi6_addr = path.local.in6().sin6_addr;
+                source_addr.v6.ipi6_ifindex = 0;
```
Already 0 via the memset

jagerman commented on src/udp.cpp:
```diff
@@ -740,8 +756,7 @@ namespace oxen::quic
             rv = sendmsg(sock_, &hdr, 0);
-            if (rv < 0)
-                break;
+            check_rv(rv, "sendmsg");
```
We shouldn't throw here.

jagerman committed:
Fix callback not passing the actual result

Sending can fail, and if it does we already have the return value on
hand: the previous code was erroneously always passing success.

jagerman committed:
log readmsg errors in debug builds; misc cleanups

- Log (in debug builds) but don't throw if readmsg fails
- misc other small cleanups
